### ISSUE TITLE
Proposal: Always get the latest buildpack from heroku

### ISSUE THREAD

**Issue by tolstenko** [teamhephy/slugbuilder#11]
### Goal:
When downloading buildpacks from heroku, always download the latest version.
### Overview
Instead of editing the file `install-buildpacks` everytime heroku creates a new version, I think it is safe to assume the new version is stable enough for us to use it. This approach will reduce maintenance burden.
If a given user wants to lock the buildpack version, it could do it by setting the BUILD_PACK env with the version:
`deis config:set BUILDPACK_URL=https://github.com/heroku/heroku-buildpack-php#v149`
The same way I had to do in order to make example-php work again for me.
I don't know when this code runs. If it runs once, this approach wont be useful. But if it runs every time, so we will always be updated with heroku builds.
### Code Changes
The script `install-buildpacks` would fetch the latest tag and use it. It be something similar to this code instead of hardcoding the version:
``` bash
#!/usr/bin/env bash
set -eo pipefail

BUILDPACK_INSTALL_PATH="/tmp/buildpacks"

download_buildpack() {
    buildpack_url="https://github.com/heroku/$1.git"
    buildpack_commit=`curl -s https://api.github.com/repos/heroku/$1/tags | python -c "import sys, json; print json.load(sys.stdin)[0]['name']"`
    buildpack_name="$2"

    echo "Fetching $buildpack_name..."

    set +e
    git clone --branch "$buildpack_commit" --depth 1 "$buildpack_url" $BUILDPACK_INSTALL_PATH/"$buildpack_name" &>/dev/null
    SHALLOW_CLONED=$?
    set -e
    if [ $SHALLOW_CLONED -ne 0 ]; then
        # if the shallow clone failed partway through, clean up and try a full clone
        rm -rf "${BUILDPACK_INSTALL_PATH:?}"/"$buildpack_name"
        git clone --quiet "$buildpack_url" $BUILDPACK_INSTALL_PATH/"$buildpack_name"
        pushd $BUILDPACK_INSTALL_PATH/"$buildpack_name" &>/dev/null
            git checkout --quiet "$buildpack_commit"
        popd &>/dev/null
    fi

    echo "Done."
}

mkdir -p $BUILDPACK_INSTALL_PATH

download_buildpack heroku-buildpack-multi          01-multi
download_buildpack heroku-buildpack-clojure        02-clojure
download_buildpack heroku-buildpack-go             03-go
download_buildpack heroku-buildpack-gradle         04-gradle
download_buildpack heroku-buildpack-grails         05-grails
download_buildpack heroku-buildpack-java           06-java
download_buildpack heroku-buildpack-nodejs         07-nodejs
download_buildpack heroku-buildpack-php            08-php
download_buildpack heroku-buildpack-play           09-play
download_buildpack heroku-buildpack-python         10-python
download_buildpack heroku-buildpack-ruby           11-ruby
download_buildpack heroku-buildpack-scala          12-scala
```
### Testing
I am running the `make tests` right now

**Comment by kingdonb** [comment]
Thank you for creating a design document!  RFC

We are generally in favor of changes that reduce the maintenance burden, although this one reduces traceability and build reproducibility.

I agree that "Heroku's latest buildpack" is probably always preferable to whatever version we have pinned, we are not maintaining any buildpacks on our own, and it's not a particularly heavy burden for users that know they need a specific buildpack version to require it themselves in a BUILDPACK_URL var.  Maybe there are some users who are not going to like this change.  Buildpack upgrades should be backwards compatible, so it would be good to have a reference supporting that.

This PR should come with a documentation update with that reference in it, if possible, and a paragraph that explains for users, how to do that (buildpack version locking with BUILDPACK_URL).

I spoke with @tolstenko offline and we agreed that in general traceability and reproducible builds are a good thing, but in this case you found that the "heroku-buildpack-php" was no longer functioning with the example, even though I recall all of the buildpack e2e tests passed when Hephy v2.20 was released.  So, something must have changed upstream and caused our build to break.

That should ideally not have been possible, but here we are.  So the facts appear to show that we do not have a purely reproducible environment now, I guess maybe somewhere in the tagged Heroku buildpacks there are upstream changes leaking in?  Or perhaps there was a release that was removed from upstreams due to security issues.  I think there is generally an expectation that you should upgrade more frequently than we are upgrading.  It would be good to improve that.

Maybe have a way that users can globally "opt-out" of buildpack latest, by pinning the configured upstream version of one or all buildpacks.  We should have a backup plan like this and document it for cluster admins, for when Heroku breaks something in the latest release and then goes on holiday, doomsday scenario where a distributed team is affected and all new builds are breaking, cluster admin needs to apply a fix across the entire workflow instance.  Maybe that's too much planning.

**Comment by Cryptophobia** [comment]
Let's all discuss. The fact that we increment buildpack versions I think is a good design because we are showing which versions we support at a minimum.

However, when things break, it's not usually our fault but the buildpack might be at fault too. Backwards compatibility with the slugs is also a factor here. 

If we do not hard-code buildpack versions then do we run the risk of breaking backwards compatibility with older slugs?

**Comment by till** [comment]
I agree, we should not auto-update the buildpacks. E.g. the Christmas release of Ruby and the release of the buildpack (on the same day) are notorious for breaking applications because people don't pin them. ;) E.g. breaking BC in patch releases.

I don't know how much other language maintainers vet their buildpacks, but I would prefer not to auto-update them as it comes back to us and doesn't allow for repeatable or immutable runs.

---

Having said all this, should hephy provide a flag (disabled by default) to auto-update? I think then it's a conscious if you turn it on/off, etc..

**Comment by duanhongyi** [comment]
Is it better to support Zip or GZ compressed package formats? just like this:

`deis config:set BUILDPACK_URL=https://github.com/user/repository/archive/branch.zip`

or

`deis config:set BUILDPACK_URL=https://my-php-buildbacks.s3.amazon.com/php.tar.gz`


I think it's more controllable and flexible.

**Comment by Cryptophobia** [comment]
> Having said all this, should hephy provide a flag (disabled by default) to auto-update? I think then it's a conscious if you turn it on/off, etc..

I like this idea. We could make it so it always pulls the latest tag on the buildpack if a specific variable is defined like `BUILDPACK_LATEST=true`. Should be an easy PR. 

Anybody, interest in making this simple PR? Should be a small change on `slugbuilder` and `controller` to add the variable and pass to slugbuilder and a new list of latest tag for the buildpack urls.

> Is it better to support Zip or GZ compressed package formats? just like this:

Could be. Have not really tested the pros vs. the cons of that approach. For one thing, you can't clearly see from the URL what version of the buildpack you are actually using which is making things more obscure and unsafe as you could be using a malicious BUILDPACK_URL unknowingly.
